common/header.go

```diff
@@ -59,18 +59,21 @@ func WithHeaderUnauthorizedHandler(
 
 // WithDynamicHeaders sets a function that will be called on every request to
 // get additional headers to use. Use this for things like time-based tokens
-// or loading headers from some external authority.
-func WithDynamicHeaders(f func() ([]Header, error)) HeaderAuthClientOption {
+// or loading headers from some external authority. The function can access a
+// copy of the request object to use its metadata for generating headers.
+func WithDynamicHeaders(f DynamicHeadersGenerator) HeaderAuthClientOption {
 	return func(params *headerClientParams) {
 		params.dynamicHeaders = f
 	}
 }
 
+type DynamicHeadersGenerator func(http.Request) ([]Header, error)
```

Comment:
I'd use `*http.Request` rather than `http.Request`, because I'm not sure what copy-by-value is going to do with certain fields (e.g. body).

Reply:
Idea was for anyone to not be able to mutate the request (accidentally or not) - it would create hard-to-debug behavior